### PR TITLE
Logging improvements

### PR DESCRIPTION
**Pull request by kichalla**
@loudej @lodejard @rynowak @Eilon

### CODE REVIEW

File: NuGet.config [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (kichalla) [inline NuGet.config]: probably BOM...will revert it

Comment (rynowak) [inline NuGet.config]: This is my favorite part so far.

File: samples/SampleApp/LoggerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        static LoggerExtensions()
+        {
+            LoggerMessage.DefineScope(out _purchaceOrderScope, "PO:{PurchaceOrder}");
+            LoggerMessage.Define(out _programStarting, LogLevel.Information, 1, "Starting", "at '{StartTime}' and 0x{Hello:X} is hex of 42");
+            LoggerMessage.Define(out _programStopping, LogLevel.Information, 2, "Stopping", "at '{StopTime}'");
+        }
+
+        public static IDisposable PurchaceOrderScope(this ILogger logger, string purchaceOrder)
+        {
+            return _purchaceOrderScope(logger, purchaceOrder);
```
Comment (rynowak) [inline samples/SampleApp/LoggerExtensions.cs]: Fix spelling here :laughing:

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+using Microsoft.Extensions.Logging.Internal;
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.Logging
+{
+    /// <summary>
+    /// Creates delegates which can be later cached to log messages in a performant way.
+    /// </summary>
+    public static class LoggerMessage
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Name seems a little iffy - @Eilon ?

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        /// <summary>
+        /// Creates a delegate which can be used for creating a log scope.
+        /// </summary>
+        /// <param name="scope">A delegate which creates a log scope</param>
+        /// <param name="formatString">The named format string</param>
+        public static void DefineScope(out Func<ILogger, IDisposable> scope, string formatString)
+        {
+            var formatter = new LogValuesFormatter(formatString);
+
+            scope = logger => logger.BeginScopeImpl(new LogValues(formatter));
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Should we close over `new LogValues` as well? In the case of just a format string we can remove that allocation

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+namespace Microsoft.Extensions.Logging
+{
+    /// <summary>
+    /// Creates delegates which can be later cached to log messages in a performant way.
+    /// </summary>
+    public static class LoggerMessage
+    {
+        /// <summary>
+        /// Creates a delegate which can be used for creating a log scope.
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: `can be used for creating a log scope` -> `can be invoked to create a log scope`

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        {
+            var formatter = new LogValuesFormatter(formatString);
+
+            scope = (logger, arg1) => logger.BeginScopeImpl(new LogValues<T1>(formatter, arg1));
+        }
+
+        /// <summary>
+        /// Creates a delegate which can be used for creating a log scope.
+        /// </summary>
+        /// <typeparam name="T1">The parameter passed to the named format string</typeparam>
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: This is the type of the parameter, not the parameter

Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Use periods, these are generally complete sentences.

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+        /// <param name="formatString">The named format string</param>
+        public static void DefineScope<T1, T2, T3>(out Func<ILogger, T1, T2, T3, IDisposable> scope, string formatString)
+        {
+            var formatter = new LogValuesFormatter(formatString);
+
+            scope = (logger, arg1, arg2, arg3) => logger.BeginScopeImpl(new LogValues<T1, T2, T3>(formatter, arg1, arg2, arg3));
+        }
+
+        /// <summary>
+        /// Creates a delegate which can be used for logging a message.
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: `can be used for logging a message` -> `can be invoked to log a message`

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
+            scope = (logger, arg1) => logger.BeginScopeImpl(new LogValues<T1>(formatter, arg1));
+        }
+
+        /// <summary>
+        /// Creates a delegate which can be used for creating a log scope.
+        /// </summary>
+        /// <typeparam name="T1">The parameter passed to the named format string</typeparam>
+        /// <typeparam name="T2">The parameter passed to the named format string</typeparam>
+        /// <param name="scope">A delegate which creates a log scope</param>
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: `delegate`?

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+        /// <summary>
+        /// Creates a delegate which can be used for logging a message.
+        /// </summary>
+        /// <typeparam name="T1">The parameter passed to the named format string</typeparam>
+        /// <param name="message">A delegate which creates a log message</param>
+        /// <param name="logLevel">The <see cref="LogLevel"/></param>
+        /// <param name="eventId">The event id</param>
+        /// <param name="eventName">The event name</param>
+        /// <param name="formatString">The named format string</param>
+        public static void Define<T1>(out Action<ILogger, T1, Exception> message, LogLevel logLevel, int eventId, string eventName, string formatString)
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Where did `eventName` from from? @loudej @lodejard

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+                }
+            };
+        }
+
+        private class LogValues : ILogValues
+        {
+            public static Func<object, Exception, string> Callback = (state, exception) => ((LogValues)state)._formatter.Format(((LogValues)state).ToArray());
+
+            private static IEnumerable<KeyValuePair<string, object>> _getValues = new KeyValuePair<string, object>[0];
+            private static object[] _toArray = new object[0];
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Find a better name for this

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+
+            private readonly LogValuesFormatter _formatter;
+
+            public LogValues(LogValuesFormatter formatter)
+            {
+                _formatter = formatter;
+            }
+
+            public IEnumerable<KeyValuePair<string, object>> GetValues() => _getValues;
```
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Is the change to make this indexable coming as part of this PR?

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs (hunk 10 above)
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Should we still include `OriginalFormat`?

File: test/Microsoft.Extensions.Logging.Test/TestLogger.cs [hunk 12]
```diff
@@ -31,7 +31,7 @@ public IDisposable BeginScopeImpl(object state)
                 Scope = state,
             });
 
-            return null;
+            return new NoopDisposable();
```
Comment (rynowak) [inline test/Microsoft.Extensions.Logging.Test/TestLogger.cs]: Why is this needed?

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 13]
```diff
@@ -0,0 +1,343 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.Logging.Internal;
```
Comment (Eilon) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Sort 'em.

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs (hunk 3 above)
Comment (Eilon) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Yeah not a huge fan of the name either. Usually these types of static classes are more verby and less nouny because they merely contain helper methods, and do not themselves represent anything.

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs (hunk 9 above)
Comment (Eilon) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: I think it's `categoryName` elsewhere: https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging/LoggerFactory.cs#L20

File: test/Microsoft.Framework.Logging.Test/LoggerMessageTest.cs [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Logging.Test;
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.Framework.Logging
+{
+    [Trait("test", "loggermessage")]
```
Comment (Eilon) [inline test/Microsoft.Framework.Logging.Test/LoggerMessageTest.cs]: Do we actually use traits anywhere?

File: test/Microsoft.Framework.Logging.Test/TestLoggerExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,110 @@
+﻿using System;
```
Comment (Eilon) [inline test/Microsoft.Framework.Logging.Test/TestLoggerExtensions.cs]: Insert file header.

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs (hunk 9 above)
Comment (kichalla) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: This would be part of the message itself and if am not wrong should ideally be different than the logger name...
In the following logger, the name of the logger is already part of the message but `eventname` would be additional detail to the message.
https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs#L79

File: test/Microsoft.Framework.Logging.Test/LoggerMessageTest.cs (hunk 14 above)
Comment (kichalla) [inline test/Microsoft.Framework.Logging.Test/LoggerMessageTest.cs]: Not in many places(ex: musicstore uses them), but I like them as it allows me to run only these specific tests...will remove it.

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+            };
+
+            public object[] ToArray() => new object[] { _value0 };
+
+            public override string ToString() => _formatter.Format(ToArray());
+        }
+
+        private class LogValues<T0, T1> : ILogValues
+        {
+            public static Func<object, Exception, string> Callback = (state, exception) => ((LogValues<T0, T1>)state)._formatter.Format(((LogValues<T0, T1>)state).ToArray());
```
Comment (kichalla) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: @lodejard @loudej Callbacks like these could just be `state.ToString()` as we are overrding ToString in them?

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+            private readonly T1 _value1;
+
+            public LogValues(LogValuesFormatter formatter, T0 value0, T1 value1)
+            {
+                _formatter = formatter;
+                _value0 = value0;
+                _value1 = value1;
+            }
+
+            public IEnumerable<KeyValuePair<string, object>> GetValues() => new[]
```
Comment (kichalla) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: @loudej @lodejard wondering why do we need to do it this way...like we could use the formatter which already does this for us `formatter.GetValues(ToArray())`
https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs#L146

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: I still have more questions than answers after reading your reply. I know what it **does** from reading the code. Why do we need it an why is it part of this PR?

File: test/Microsoft.Extensions.Logging.Test/TestLogger.cs (hunk 12 above)
Comment (kichalla) [inline test/Microsoft.Extensions.Logging.Test/TestLogger.cs]: This was to make sure that the test logger's scope was created when the func was executed

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs (hunk 10 above)
Comment (kichalla) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: :+1: looks like `OriginalFormat` should be included based on how `LogValuesFormatter` implements this. Updated the test accordingly.

**Comment by kichalla** [comment]
Updated @loudej @lodejard @rynowak @Eilon

**Comment by rynowak** [comment]
:shipit:

**Comment by kichalla** [comment]
7c01b9c3639801d89bd88bb607600c57dfd4cabf
